keps/sig-architecture/2558-field-version-openapi/README.md

```diff
@@ -0,0 +1,414 @@
+
+N/A
+
+###### What specific metrics should inform a rollback?
+
+N/A
+
+###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?
+
+N/A
+
+###### Is the rollout accompanied by any deprecations and/or removals of features, APIs, fields of API types, flags, etc.?
+
+N/A
+
+### Monitoring Requirements
+
+N/A
+
+###### How can an operator determine if the feature is in use by workloads?
+
+N/A
+
+###### What are the SLIs (Service Level Indicators) an operator can use to determine the health of the service?
+
+N/A
+
+###### What are the reasonable SLOs (Service Level Objectives) for the above SLIs?
+
+N/A
+
+###### Are there any missing metrics that would be useful to have to improve observability of this feature?
+
+N/A
+
+### Dependencies
+
+N/A
+
+###### Does this feature depend on any specific services running in the cluster?
+
+N/A
+
+### Scalability
+
+N/A
+
+###### Will enabling / using this feature result in any new API calls?
+
+N/A
+
+###### Will enabling / using this feature result in introducing new API types?
+
+N/A
+
+###### Will enabling / using this feature result in any new calls to the cloud provider?
+
+N/A
+
+###### Will enabling / using this feature result in increasing size or count of the existing API objects?
+
+N/A
+
+###### Will enabling / using this feature result in increasing time taken by any operations covered by existing SLIs/SLOs?
+
+N/A
+
+###### Will enabling / using this feature result in non-negligible increase of resource usage (CPU, RAM, disk, IO, ...) in any components?
+
+N/A
+
+### Troubleshooting
+
+N/A
+
+###### How does this feature react if the API server and/or etcd is unavailable?
+
+N/A
+
+###### What are other known failure modes?
+
+N/A
+
+###### What steps should be taken if SLOs are not being met to determine the problem?
+
+N/A
+
+## Implementation History
+
+- 2021-02-05: Initial [discussion on the SIG Architecture mailing list]
+- 2021-03-07: KEP proposed
+
+## Alternatives
+
+1. The [prerelease-lifecycle-gen] generator allows specifying metadata about
+the prerelease status for packages (e.g. `cronjobs.v1beta1.batch`).
+
+Field-level versioning information support is added in the kube-openapi
+generator and not added to the [prerelease-lifecycle-gen] generator
+because:
+
+- [prerelease-lifecycle-gen] only generates and documents metadata
+about when each _package_ was introduced, deprecated or removed.
+It is not meant to work for individual API types or fields.
+- Additionally, even if field-level support was added to it,
+this information would not be surfaced in OpenAPI which makes it
+harder to be consumed by other tools (e.g. for conformance and documentation).
+
+2. The initial draft of this KEP allowed specifying multiple prerelease statuses
+in comment tags for an API field. For example:
+
+```
+// +k8s:openapi-gen:prerelease=alpha,minVersion=v1.20,featureGate=Frobber2D
+// +k8s:openapi-gen:prerelease=beta,minVersion=v1.21,featureGate=Frobber2D
```

Comment:
From [slack](https://kubernetes.slack.com/archives/C5P3FE08M/p1615126034021600?thread_ts=1615058923.018400&cid=C5P3FE08M), @sftim pointed out that documenting the entire graduation history might be useful for SIG Docs to generate https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/